src/raw/mod.rs

```diff
@@ -1442,6 +1442,40 @@ impl RawTableInner {
     }
 }
 
+/// Find the previous power of 2. If it's already a power of 2, it's unchanged.
+/// Passing zero is undefined behavior.
+pub(crate) fn prev_pow2(z: usize) -> usize {
+    let shift = mem::size_of::<usize>() * 8 - 1;
+    1 << (shift - (z.leading_zeros() as usize))
+}
+
+fn maximum_buckets_in(
+    allocation_size: usize,
+    table_layout: TableLayout,
+    group_width: usize,
+) -> usize {
+    // Given an equation like:
+    //   z >= x * y + x + g
+    // x can be maximized by doing:
+    //   x = (z - g) / (y + 1)
+    // If you squint:
+    //   x is the number of buckets
+    //   y is the table_layout.size
+    //   z is the size of the allocation
+    //   g is the group width
+    // But this is ignoring the padding needed for ctrl_align.
+    // If we remember these restrictions:
+    //   x is always a power of 2
+    //   Layout size for T must always be a multiple of T
+    // Then the alignment can be ignored if we add the constraint:
+    //   x * y >= table_layout.ctrl_align
+    // This is taken care of by `capacity_to_buckets`.
```

Comment:
This is brittle: currently `capacity_to_buckets` doesn't actually guarantee this. If we are going to rely on this guarantee then `capacity_to_buckets` needs a `debug_assert` and a comment indicating that `maximum_buckets_in` relies on this behavior. This will remind anyone who changes `capacity_to_buckets` in the future to also adjust `maximum_buckets_in`.

Reply:
Since #615, it _does_ guarantee this. Well, assuming I didn't make a mistake anyway ^_^. I tried finding issues with it but I don't see any, let me know if you see something.

I think adding assertions can be helpful to make this more obvious and be more resilient to future changes, so I've added `ensure_bucket_bytes_at_least_ctrl_align`. Take a look.